pkg/master/master.go

```diff
@@ -910,94 +765,155 @@ func (m *Master) getExtensionResources(c *Config) map[string]rest.Storage {
+		apiGroupInfo.VersionedResourcesStorageMap[policyapiv1alpha1.SchemeGroupVersion.Version] = storageForVersion(policyapiv1alpha1.SchemeGroupVersion)
+		apiGroupInfo.GroupMeta.GroupVersion = policyapiv1alpha1.SchemeGroupVersion
 	}
-	if c.APIResourceConfigSource.ResourceEnabled(version.WithResource("rolebindings")) {
-		roleBindingsStorage := rolebindingetcd.NewREST(m.GetRESTOptionsOrDie(c, rbac.Resource("rolebindings")))
-		storage["rolebindings"] = rolebindingpolicybased.NewStorage(roleBindingsStorage, newRuleValidator(), c.AuthorizerRBACSuperUser)
+
+	return apiGroupInfo, true
+}
+
+func buildAppsResources(apiResourceConfigSource genericapiserver.APIResourceConfigSource, restOptionsGetter RESTOptionsGetter) (genericapiserver.APIGroupInfo, bool) {
+	apiGroupInfo := NewDefaultAPIGroupInfo(appsapi.GroupName)
+
+	storageForVersion := func(version unversioned.GroupVersion) map[string]rest.Storage {
+		storage := map[string]rest.Storage{}
+		if apiResourceConfigSource.ResourceEnabled(version.WithResource("petsets")) {
+			petsetStorage, petsetStatusStorage := petsetetcd.NewREST(restOptionsGetter(apps.Resource("petsets")))
+			storage["petsets"] = petsetStorage
+			storage["petsets/status"] = petsetStatusStorage
+		}
+		return storage
 	}
-	if c.APIResourceConfigSource.ResourceEnabled(version.WithResource("clusterroles")) {
-		clusterRolesStorage := clusterroleetcd.NewREST(m.GetRESTOptionsOrDie(c, rbac.Resource("clusterroles")))
-		storage["clusterroles"] = clusterrolepolicybased.NewStorage(clusterRolesStorage, newRuleValidator(), c.AuthorizerRBACSuperUser)
+
+	if apiResourceConfigSource.AnyResourcesForVersionEnabled(appsapi.SchemeGroupVersion) {
+		apiGroupInfo.VersionedResourcesStorageMap[appsapi.SchemeGroupVersion.Version] = storageForVersion(appsapi.SchemeGroupVersion)
+		apiGroupInfo.GroupMeta.GroupVersion = appsapi.SchemeGroupVersion
 	}
-	if c.APIResourceConfigSource.ResourceEnabled(version.WithResource("clusterrolebindings")) {
-		clusterRoleBindingsStorage := clusterrolebindingetcd.NewREST(m.GetRESTOptionsOrDie(c, rbac.Resource("clusterrolebindings")))
-		storage["clusterrolebindings"] = clusterrolebindingpolicybased.NewStorage(clusterRoleBindingsStorage, newRuleValidator(), c.AuthorizerRBACSuperUser)
+
+	return apiGroupInfo, true
+}
+
+func buildRBACResources(authorizerRBACSuperUser string) RESTStorageProvider {
+	return func(apiResourceConfigSource genericapiserver.APIResourceConfigSource, restOptionsGetter RESTOptionsGetter) (genericapiserver.APIGroupInfo, bool) {
+		apiGroupInfo := NewDefaultAPIGroupInfo(rbac.GroupName)
+
+		storageForVersion := func(version unversioned.GroupVersion) map[string]rest.Storage {
+			once := new(sync.Once)
+			var authorizationRuleResolver rbacvalidation.AuthorizationRuleResolver
+			newRuleValidator := func() rbacvalidation.AuthorizationRuleResolver {
+				once.Do(func() {
+					authorizationRuleResolver = rbacvalidation.NewDefaultRuleResolver(
```

Comment:
> This whole function looks like word salad on a narrow screen (i.e., github side by side diff). What do you think about naming some parts of this and promoting them to be top level so there's not so much indentation? Would be helpful for the other functions too, but I think this is the most extreme.

I agree.  Mind if I do that as a step two with the `RESTStorageProviderFunc` refactor?  I can do it one go, but I think the diff gets even harder.  I'd promise it as a quick followup, less than a week.